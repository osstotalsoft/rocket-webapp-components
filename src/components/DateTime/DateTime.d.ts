import { ParsableDate } from "@material-ui/pickers/constants/prop-types";

import { KeyboardDateTimePickerProps } from "@material-ui/pickers";

export interface DateTimeProps extends Omit<KeyboardDateTimePickerProps, 'onChange' | 'value'> {
    /**
    * Picker value.
    * @default null
    */
    value?: ParsableDate
    /**
    * @default () => { }
    * onChange callback.
    */
    onChange?: (date: DateIOType) => void
    /**
    * @default  'HH:mm'
    * Date format string.
    */
    dateFormat?: string
    /**
    * @default  'DD-MM-YYYY'
    * Time format string.
    */
    timeFormat?: string
    /**
    * @default false
    * If true, the picker and text field will be disabled.
    */
    disabled?: boolean
    /**
    * @defualt 'inline'
    * Picker container option.
    */
    variant?: "dialog" | "inline" | "static"
    /**
    * @default ["date"]
    * Array of views to show
    */
    views?: ("date" | "year" | "month" | "hours" | "minutes")[]
    /**
    * @default false
    * If true, the time is shown.
    */
    showTime?: boolean
    /**
    * @default false
    * If true, the error data is provided.
    */
    error?: boolean
    /**
    * The text content in case of failure.
    */
    helperText?: string
    /**
    * If true, show clear action in picker dialog.
    */
    clearable?: boolean
    /**
    * @default 'Invalid Date Format'
    * Message, appearing when date cannot be parsed.
    */
    invalidDateMessage?: React.ReactNode
    /**
    * @default 'Date should not be after maximal date'
    * Error message, shown if date is more then maximal date. 
    */
    maxDateMessage?: React.ReactNode
    /**
    * @default 'Date should not be before minimal date'
    * Error message, shown if date is less then minimal date
    */
    minDateMessage?: React.ReactNode
    /**
     * Disable past dates
     */
    disablePast?: boolean
    /**
     * Disable future dates
     */
    disableFuture?: boolean
    /**
    * @default "en"
    * To use another localization import "moment/locale/[lang]" in your app
    * The current language, preferably taken from the i18next (i18.language) or another internationalization library
    */
    language?: string
    /**
    * Hide toolbar and show only date/time views
    * @default false
    */
    disableToolbar: boolean
    /**
    * @default false
    * Auto accept date on selection
    */
    autoOk: boolean
    /**
    * @default true
    * 12h/24h view for hour selection clock
    */
    ampm: boolean
}
/**
 *
 * Demos:
 *
 * - https://bit.dev/totalsoft_oss/react-mui/date-time
 *
 *  Take a look over the code for further information:https://github.com/osstotalsoft/rocket-webapp-components/blob/master/components/DateTime/DateTime.js
 * 
 * Material-UI component used:
 * 
 * - KeyboardDateTimePicker: https://material-ui-pickers.dev/api/KeyboardDateTimePicker
 * 
 */
export default function DateTime(props: DateTimeProps): JSX.Element;