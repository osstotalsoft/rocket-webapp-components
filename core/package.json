{
  "engines": {
    "node": ">=10"
  },
  "scripts": {
    "prebuild": "rimraf build",
    "build": "npm run build:cjs &&  npm run build:copy-files",
    "build:cjs": "babel src -d build --copy-files",
    "build:copy-files": "node ./scripts/copy-files.js",
    "bit-build": "bit -v && bit import && bit build"
  },
  "devDependencies": {
    "@babel/cli": "^7.8.0",
    "@babel/core": "^7.7.7",
    "@babel/node": "^7.7.7",
    "@babel/plugin-proposal-decorators": "^7.8.3",
    "@babel/preset-env": "^7.7.7",
    "@babel/preset-react": "^7.8.0",
    "babel-eslint": "^10.0.3",
    "cross-env": "^7.0.2",
    "eslint": "^6.8.0",
    "eslint-config-prettier": "^6.7.0",
    "eslint-config-standard": "^14.1.0",
    "eslint-config-standard-react": "^9.2.0",
    "eslint-plugin-import": "^2.18.2",
    "eslint-plugin-node": "^11.0.0",
    "eslint-plugin-prettier": "^3.1.1",
    "eslint-plugin-promise": "^4.2.1",
    "eslint-plugin-react": "^7.17.0",
    "eslint-plugin-standard": "^4.0.1",
    "gh-pages": "^2.2.0",
    "npm-run-all": "^4.1.5",
    "prettier": "^1.19.1",
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "react-scripts": "^3.4.1",
    "rimraf": "^3.0.0",
    "url-search-params-polyfill": "^7.0.1"
  },
  "peerDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "react-flags": "^0.1.18",
    "react-router-dom": "^5.1.2"
  },
  "sideEffects": false,
  "dependencies": {},
  "bit": {
    "env": {
      "compiler": "totalsoft_oss.react-mui/compilers/react@1.0.32"
    },
    "componentsDefaultDirectory": "components/{name}",
    "packageManager": "npm"
  }
}
