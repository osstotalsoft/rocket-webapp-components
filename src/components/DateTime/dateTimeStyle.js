import styles from '../../assets/jss/styles';

const dateTimeStyles = (theme) => {
  const { blackColor, blueColor, defaultFont } = styles(theme);

  return ({
    dateTimeInput: {
      ...defaultFont,
      minWidth: "100%",
      display: 'flex'
    },
    label: {
      ...defaultFont,
      fontWeight: "400",
      display: 'flex'
    },
    action: {
      cursor: 'pointer'
    },
    dateTimeIconButtons: {
      ...defaultFont,
      padding: '0px',
      color: blackColor
    },
    themeColor: blueColor,
    defaultFont
  })
}

export default dateTimeStyles;