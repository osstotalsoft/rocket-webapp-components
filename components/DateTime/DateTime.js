import React, { useCallback, useState } from "react";
import PropTypes from "prop-types";
import { makeStyles, IconButton } from "@material-ui/core";
import { Clear, CalendarToday } from "@material-ui/icons";

import {
  MuiPickersUtilsProvider,
  KeyboardDateTimePicker
} from "@material-ui/pickers";
import MomentUtils from "@date-io/moment";

import dateTimeStyle from "./dateTimeStyle";
import moment from "moment";
import { emptyObject } from "../../utils/constants";

// ! To use another localization import "moment/locale/[lang]" in your app
moment.locale("en");

const useStyles = makeStyles(dateTimeStyle);
const defaultTimeFormat = "HH:mm";
const defaultDateFormat = "DD-MM-YYYY";

const TooltipArea = React.forwardRef(({ children, ...props }, ref) => (
  <div {...props} ref={ref}>
    {children}
  </div>
));
TooltipArea.displayName = "TooltipArea";
TooltipArea.propTypes = {
  children: PropTypes.element
};

function DateTime({
  value,
  onChange,
  dateFormat,
  timeFormat,
  showTime,
  clearable,
  disabled,
  error,
  helperText,
  invalidDateMessage,
  minDateMessage,
  maxDateMessage,
  language,
  ...rest
}) {
  const classes = useStyles();
  const [open, setOpen] = useState(false);
  const format = showTime ? `${dateFormat} ${timeFormat}` : dateFormat;

  const errorData = disabled
    ? { error: false, helperText: null }
    : error
    ? { error, helperText }
    : emptyObject;

  const handleSetOpen = useCallback(value => () => setOpen(value), []);
  const handleChange = useCallback(
    date => onChange(date ? moment(date).toDate() : null),
    [onChange]
  );
  const handleClear = useCallback(() => handleChange(null), [handleChange]);

  const clearablePickerProps = {
    InputProps: {
      endAdornment: (
        <>
          <IconButton
            disabled={disabled}
            onClick={handleClear}
            className={classes.dateTimeIconButtons}
          >
            <Clear fontSize="small" />
          </IconButton>
          <IconButton
            disabled={disabled}
            className={classes.dateTimeIconButtons}
            onClick={handleSetOpen(true)}
          >
            <CalendarToday fontSize="small" />
          </IconButton>
        </>
      )
    },
    InputAdornmentProps: {
      position: "start",
      style: {
        display: "none"
      }
    }
  };

  const simplePickerProps = {
    keyboardIcon: <CalendarToday fontSize='small' onClick={handleSetOpen(true)} />,
    KeyboardButtonProps: {
      disabled: disabled,
      className: classes.dateTimeIconButtons
    }
  }

  const pickerProps =
    clearable && value ? clearablePickerProps : simplePickerProps;

  return (
    <MuiPickersUtilsProvider
      libInstance={moment}
      utils={MomentUtils}
      locale={language}
    >
      <KeyboardDateTimePicker
        {...pickerProps}
        fullWidth
        value={value}
        open={open}
        onClose={handleSetOpen(false)}
        onChange={handleChange}
        format={format}
        invalidDateMessage={invalidDateMessage}
        minDateMessage={minDateMessage}
        maxDateMessage={maxDateMessage}
        className={classes.defaultFont}
        InputProps={{
          disabled,
          className: classes.dateTimeInput,
          ...pickerProps.InputProps
        }}
        InputLabelProps={{
          className: classes.label
        }}
        {...errorData}
        {...rest}
      />
    </MuiPickersUtilsProvider>
  );
}

DateTime.defaultProps = {
  value: null,
  onChange: () => {},
  dateFormat: defaultDateFormat,
  timeFormat: defaultTimeFormat,
  variant: "inline",
  views: ["date"],
  disabled: false,
  invalidDateMessage: "Invalid Date Format",
  minDateMessage: "Date should not be before minimal date",
  maxDateMessage: "Date should not be after maximal date",
  disableToolbar: true,
  autoOk: true,
  ampm: false
};

DateTime.propTypes = {
  value: PropTypes.any,
  onChange: PropTypes.func,
  dateFormat: PropTypes.string,
  timeFormat: PropTypes.string,
  disabled: PropTypes.bool,
  variant: PropTypes.oneOf(["dialog", "inline", "static"]),
  views: PropTypes.arrayOf(
    PropTypes.oneOf(["date", "year", "month", "hours", "minutes"])
  ),
  showTime: PropTypes.bool,
  error: PropTypes.bool,
  helperText: PropTypes.string,
  clearable: PropTypes.bool,
  invalidDateMessage: PropTypes.node,
  maxDateMessage: PropTypes.node,
  minDateMessage: PropTypes.node,
  language: PropTypes.string,
  disableToolbar: PropTypes.bool,
  autoOk: PropTypes.bool,
  ampm: PropTypes.bool
};

export default DateTime;
