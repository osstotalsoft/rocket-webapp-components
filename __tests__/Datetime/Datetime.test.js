import React from "react";
import renderer from "react-test-renderer";
import DateTime from "../../components/DateTime/DateTime";
import { mount } from "enzyme";
import { IconButton } from "@material-ui/core";
import { CalendarToday } from "@material-ui/icons";
import { KeyboardDateTimePicker } from "@material-ui/pickers";

describe("DateTime", () => {
  it("snapshot", () => {
    const onDateTimeChanged = () => {};
    const onHandleClick = () => {};
    const defaultTimeFormat = "HH:mm";

    const component = renderer.create(
      <DateTime
        value={"20190101"}
        label={"Datetime"}
        onChange={onDateTimeChanged}
        onClick={onHandleClick}
        dateFormat="DD-MMMM-YYYY"
        timeFormat={defaultTimeFormat}
        clearable={true}
      />
    );

    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it("if clearable and value are true then InputProps node will be made up of two IconButtons", () => {
    const onDateTimeChanged = jest.fn();
    const onHandleClick = jest.fn();
    const defaultTimeFormat = "HH:mm";

    const wrapper = mount(
      <DateTime
        value={"value"}
        label={"Datetime"}
        onChange={onDateTimeChanged}
        dateFormat="DD-MMMM-YYYY"
        timeFormat={defaultTimeFormat}
        clearable={true}
        onClick={onHandleClick}
      />
    );

    expect(wrapper.find(IconButton)).toHaveLength(3);
  });

  it("if clearable and value are false then InputProps node will be made of one CalendarToday node", () => {
    const onDateTimeChanged = () => {};
    const defaultTimeFormat = "HH:mm";

    const wrapper = mount(
      <DateTime
        value={null}
        label={"Datetime"}
        onChange={onDateTimeChanged}
        dateFormat="DD-MMMM-YYYY"
        timeFormat={defaultTimeFormat}
        clearable={false}
      />
    );

    expect(wrapper.find(CalendarToday)).toHaveLength(1);
  });

  it("classes defaultFont and dateTimeInput should be present", () => {
    const onDateTimeChanged = jest.fn();
    const defaultTimeFormat = "HH:mm";

    const wrapper = mount(
      <DateTime
        value={null}
        label={"Datetime"}
        onChange={onDateTimeChanged}
        dateFormat="DD-MMMM-YYYY"
        timeFormat={defaultTimeFormat}
        clearable={false}
      />
    );

    expect(wrapper.find(KeyboardDateTimePicker).props().className).toContain(
      "defaultFont"
    );
    expect(
      wrapper.find(KeyboardDateTimePicker).props().InputProps.className
    ).toContain("dateTimeInput");
  });
});
