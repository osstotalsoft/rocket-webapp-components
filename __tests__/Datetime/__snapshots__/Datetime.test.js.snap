// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTime snapshot 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root makeStyles-defaultFont-6 MuiFormControl-fullWidth"
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root makeStyles-label-2 MuiInputLabel-formControl MuiInputLabel-animated"
    data-shrink={false}
  >
    Datetime
  </label>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-dateTimeInput-1 MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
      disabled={false}
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
      value=""
    />
    <div
      className="MuiInputAdornment-root MuiInputAdornment-positionEnd"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root makeStyles-dateTimeIconButtons-4"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
            focusable="false"
            onClick={[Function]}
            viewBox="0 0 24 24"
          >
            <path
              d="M20 3h-1V1h-2v2H7V1H5v2H4c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 18H4V8h16v13z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  
</div>
`;
